contracts/hydro/src/contract.rs

```diff
@@ -2753,6 +2771,128 @@ pub fn convert_lockup_to_dtoken_reply(
+            .validate_denom(&deps.as_ref(), current_round, coin.denom.clone())
+            .map_err(|err| {
+                new_generic_error(format!("Error validating denom '{}': {}", coin.denom, err))
+            })?;
+        let ratio = token_manager
+            .get_token_group_ratio(&deps.as_ref(), current_round, token_group_id)
+            .map_err(|err| {
+                new_generic_error(format!(
+                    "Error fetching token group ratio for '{}': {}",
+                    coin.denom, err
+                ))
+            })?;
+        let coin_as_base = Decimal::from_atomics(coin.amount.u128(), 18).unwrap() * ratio;
+        available_base_atom += coin_as_base.atomics().u128();
+        fund_conversion.push((coin.denom.clone(), coin.amount.u128(), ratio));
+    }
+
+    // Step 4: Determine how much to deduct and update state
+    let full_payment = available_base_atom >= slash_base_atom.atomics().u128();
+    if full_payment {
+        LOCKS_PENDING_SLASHES.remove(deps.storage, lockup.lock_id);
+    } else {
+        let remaining_base =
+            slash_base_atom - Decimal::from_atomics(available_base_atom, 18).unwrap();
+        let remaining_original = remaining_base / slash_ratio;
+        LOCKS_PENDING_SLASHES.save(
+            deps.storage,
+            lockup.lock_id,
+            &Uint128::from(remaining_original.atomics().u128()),
+        )?;
+    }
+
+    // Step 5: Figure out what was used and what should be refunded
+    let mut funds_used: HashMap<String, u128> = HashMap::new();
+    let mut remaining_atom = core::cmp::min(slash_base_atom.atomics().u128(), available_base_atom);
+
+    for (denom, original_amount, ratio) in &fund_conversion {
+        if remaining_atom == 0 {
+            break;
+        }
+        let coin_as_atom = Decimal::from_atomics(*original_amount, 18).unwrap() * *ratio;
+        let take_atom = core::cmp::min(coin_as_atom.atomics().u128(), remaining_atom);
+        let take_original = Decimal::from_atomics(Uint128::from(take_atom), 18).unwrap() / *ratio;
+
+        funds_used.insert(denom.clone(), take_original.atomics().u128());
+        remaining_atom -= take_atom;
+    }
+
+    // Step 6: Prepare refund messages
+    let mut refund_msgs: Vec<BankMsg> = vec![];
+    for (denom, original_amount, _) in &fund_conversion {
+        let used = funds_used.get(denom).copied().unwrap_or(0);
+        let total = *original_amount;
+        if total > used {
+            let refund_amount = total - used;
+            if refund_amount > 0 {
+                refund_msgs.push(BankMsg::Send {
+                    to_address: info.sender.to_string(),
+                    amount: vec![Coin {
+                        denom: denom.clone(),
+                        amount: Uint128::from(refund_amount),
+                    }],
+                });
+            }
+        }
+    }
+
+    Ok(Response::new()
+        .add_attribute("action", "buyout_pending_slash")
```

Comment:
Add the following attributes: sender, lock_id, coins spent to buyout the pending slash.